crates/net/rpc/src/engine/jwt_secret.rs

```diff
@@ -0,0 +1,252 @@
+use std::time::{SystemTime, UNIX_EPOCH};
+
+use hex::encode as hex_encode;
+use jsonwebtoken::{decode, errors::ErrorKind, Algorithm, DecodingKey, Validation};
+use rand::Rng;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+/// Errors returned by the [`JwtSecret`][crate::engine::JwtSecret]
+/// and the [`Jwt`][crate::engine::JwtValidator] structs.
+#[derive(Error, Debug)]
+#[allow(missing_docs)]
+pub enum JwtError {
+    #[error(transparent)]
+    JwtSecretHexDecodeError(#[from] hex::FromHexError),
+    #[error("JWT key is expected to have a length of {0} digits. {1} digits key provided.")]
+    InvalidLength(usize, usize),
+    #[error("Unsupported signature algorithm. Only HS256 is supported")]
+    UnsupportedSignatureAlgorithm,
+    #[error("The provided signature is invalid.")]
+    InvalidSignature,
+    #[error("The iat (issued-at) claim is not within +-60 seconds from the current time")]
+    InvalidIssuanceTimestamp,
+    #[error("JWT decoding error {0}")]
+    JwtDecodingError(String),
+}
+
+/// Length of the hex-encoded 256 bit secret key.
+/// A 256-bit encoded string in Rust has a length of 64 digits because each digit represents 4 bits
+/// of data. In hexadecimal representation, each digit can have 16 possible values (0-9 and A-F), so
+/// 4 bits can be represented using a single hex digit. Therefore, to represent a 256-bit string,
+/// we need 64 hexadecimal digits (256 bits ÷ 4 bits per digit = 64 digits).
+const JWT_SECRET_LEN: usize = 64;
+
+/// The JWT `iat` (issued-at) claim cannot exceed +-60 seconds from the current time.
+const JWT_MAX_IAT_DIFF: u64 = 60;
+
+/// The execution layer client MUST support at least the following alg HMAC + SHA256 (HS256)
+const JWT_SIGNATURE_ALGO: Algorithm = Algorithm::HS256;
+
+/// Value-object holding a reference to an hex-encoded 256 bit secret key.
+/// A JWT secret key is used to secure JWT-based authentication. The secret key is
+/// a shared secret between the server and the clienta and is used to calculate a digital signature
+/// for the JWT, which is included in the JWT along with its payload.
+/// -----------------------------------------------------------------
+/// [Secret key - Engine API specs](https://github.com/ethereum/execution-apis/blob/main/src/engine/authentication.md#key-distribution)
+#[derive(Debug)]
+pub struct JwtSecret(String);
+
+impl JwtSecret {
+    /// Creates an instance of [`JwtSecret`][crate::engine::JwtSecret].
+    /// Generates and error if one of the following applies:
+    /// - The `hex` argument is not a valid hexadecimal string
+    /// - The `hex` argument length is less than `JWT_SECRET_LEN`
+    pub fn new(hex: String) -> Result<JwtSecret, JwtError> {
```

Comment:
this is essentially a `from_hex` function, right?

suggest renaming

Reply:
Or just `TryFrom<String>` and make it `JwtSecret(Bytes)` or something.

Reply:
I implemented the TryInto trait, it looks nicer.
I also builds `JwtSecret` on top fo bytes